pkg/cloud/cloud.go

```diff
@@ -835,6 +836,57 @@ func (c *cloud) getVolume(ctx context.Context, request *ec2.DescribeVolumesInput
 	return volumes[0], nil
 }
 
+func (c *cloud) GetVolumeAttachLimit(ctx context.Context, nodeID string) (int, error) {
+	instance, err := c.getInstance(ctx, nodeID)
+	if err != nil {
+		return 0, err
+	}
+
+	instanceType := string(*instance.InstanceType)
+	networkInterfacesAttached := len(instance.NetworkInterfaces)
+	blockDevicesAttached := len(instance.BlockDeviceMappings)
```

Comment:
I believe (might need to verify) that nodegetinfo gets called once at driver startup when it registers with kubelet and then never again afterward. If this `blockDevicesAttached`  counts volumes attached by the driver itself, it could result inthe limit unexpectedly lowering in case the driver is restarted (e.g. if I upgrade it). For example this would cut the limit effectively in half because if the instance type has limit 10 , then it attaches 5 volumes, then I restart the driver (let's say I'm upgrading it) then this function will return 10-5 = 5 so it will refuse to attach more.

Reply:
Yes, you are right. NodeGetInfo is called once at driver startup. Like you said, if the driver is restarted, then it would reduce the number of attachments to its current number. Would you rather suggest a different approach to it?

Reply:
we would need a way to distinguish volumes attached by csi, which should not affect the limit, from volumes that are not, which should affect the limit, and I don't think describeinstances has enough info for that. I think we would need to call describeVolumes once with all the volumes and filter out the csi-specific tags? But that is another API call and I am not sure we should be making an API call in the first place.